File: android/app/src/main/java/com/clingsync/android/RecentTargets.kt
package com.clingsync.android

import android.content.Context
import org.json.JSONArray

// Persists the destination paths most recently used for a share upload (most
// recent first, de-duplicated, capped), offered by the share screen alongside the
// configured settings prefix.
object RecentTargets {
    const val MAX = 10
    private const val PREFS = "recent_share_targets"
    private const val KEY = "targets"

    fun load(context: Context): List<String> {
        val raw = prefs(context).getString(KEY, null) ?: return emptyList()
        val array = JSONArray(raw)
        return List(array.length()) { array.getString(it) }
    }

    fun record(
        context: Context,
        target: String,
    ) {
        val normalized = normalizeTarget(target)
        val updated = (listOf(normalized) + load(context).filter { it != normalized }).take(MAX)
        val array = JSONArray()
        updated.forEach { array.put(it) }
        prefs(context).edit().putString(KEY, array.toString()).apply()
    }

    private fun prefs(context: Context) = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
}

// The repo path the upload actually uses trims surrounding slashes, so normalize
// targets the same way before storing or comparing them, or equivalent paths
// ("photos", "/photos/") would show up as separate recent entries.
fun normalizeTarget(target: String): String = target.trim().trim('/')

// The target-dir choices the share screen offers: the recent targets (most recent
// first) plus the settings prefix when not already present, with the default being
// the most recently used or the settings prefix when there are none. Pure so the
// option logic is assertable on its own.
data class ShareTargetOptions(
    val options: List<String>,
    val default: String,
) {
    companion object {
        fun from(
            settingsPrefix: String,
            recent: List<String>,
        ): ShareTargetOptions {
            val prefix = normalizeTarget(settingsPrefix)
            val options = if (recent.contains(prefix)) recent else recent + prefix
            return ShareTargetOptions(options, recent.firstOrNull() ?: prefix)
        }
    }
}

File: ios/ClingSync/RecentTargets.swift
import Foundation

// Persists the destination paths most recently used for a share upload (most
// recent first, de-duplicated, capped), so the share screen can offer them as
// quick targets alongside the configured settings prefix.
enum RecentTargets {
    private static let key = "recentShareTargets"
    static let maxCount = 10

    static func load(_ defaults: UserDefaults = .standard) -> [String] {
        defaults.stringArray(forKey: key) ?? []
    }

    static func record(_ target: String, _ defaults: UserDefaults = .standard) {
        let normalized = normalizeTarget(target)
        var list = load(defaults).filter { $0 != normalized }
        list.insert(normalized, at: 0)
        defaults.set(Array(list.prefix(maxCount)), forKey: key)
    }
}

// The repo path the upload actually uses trims surrounding slashes, so normalize
// targets the same way before storing or comparing them, or equivalent paths
// ("photos", "/photos/") would show up as separate recent entries.
func normalizeTarget(_ target: String) -> String {
    target
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
}

// The target-dir choices the share screen offers: the recent targets (most recent
// first) plus the settings prefix when it is not already among them. The default
// is the most recently used, falling back to the settings prefix when there are
// none. Pure so the share store's option logic is assertable on its own.
struct ShareTargetOptions: Equatable {
    let options: [String]
    let defaultTarget: String

    init(settingsPrefix: String, recent: [String]) {
        let prefix = normalizeTarget(settingsPrefix)
        var ordered = recent
        if !ordered.contains(prefix) {
            ordered.append(prefix)
        }
        options = ordered
        defaultTarget = recent.first ?? prefix
    }
}
